### PR TITLE
nautilus: cephfs: client: add ability to lookup snapped inodes by inode number

### PR DESCRIPTION
**Pull request by smithfarm**
backport trackers:

* https://tracker.ceph.com/issues/49514
* https://tracker.ceph.com/issues/50290
* https://tracker.ceph.com/issues/50625

---

backport of

* https://github.com/ceph/ceph/pull/39294
* https://github.com/ceph/ceph/pull/40642
* https://github.com/ceph/ceph/pull/40903

parent trackers: 

* https://tracker.ceph.com/issues/48991
* https://tracker.ceph.com/issues/50112
* https://tracker.ceph.com/issues/50216

this backport was staged using ceph-backport.sh version 16.0.0.6848
find the latest version at https://github.com/ceph/ceph/blob/master/src/script/ceph-backport.sh

### CODE REVIEW

**Comment by yuriw** [comment]
@smithfarm pls rebase

**Comment by smithfarm** [comment]
@yuriw Rebased

**Comment by batrick** [comment]
blocked by fix for https://tracker.ceph.com/issues/50216#change-192030

cc @lxbsz

**Comment by smithfarm** [comment]
> blocked by fix for https://tracker.ceph.com/issues/50216#change-192030

Fix included. DNM removed. Optimistically (and possibly unreasonably) expecting a smooth make check experience.

**Comment by lxbsz** [comment]
```
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/mds/MDSDaemon.cc:45:
/home/jenkins-build/build/workspace/ceph-pull-requests/src/mds/SnapServer.h: In member function 'void SnapServer::mark_base_recursively_scrubbed(inodeno_t)':
/home/jenkins-build/build/workspace/ceph-pull-requests/src/mds/SnapServer.h:122:17: error: 'MDS_INO_ROOT' was not declared in this scope; did you mean 'CEPH_INO_ROOT'?
  122 |     if (ino ==  MDS_INO_ROOT)
      |                 ^~~~~~~~~~~~
      |                 CEPH_INO_ROOT

```
The same here, you  need `s/MDS_INO_ROOT/CEPH_INO_ROOT/` in `MDSDaemon.cc`.

**Review by lxbsz — APPROVED** [review]

**Comment by yuriw** [comment]
@smithfarm this seem to be breaking the build

https://shaman.ceph.com/builds/ceph/wip-yuri5-testing-2021-05-19-0809-nautilus/

**Comment by smithfarm** [comment]
Thanks, @yuriw - this affects the octopus backport as well: the problem is being discussed there.

**Comment by yuriw** [comment]
@smithfarm I suspect this PR beaks the build

`build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/gigantic/release/14.2.21-260-g73cf0b96/rpm/el7/BUILD/ceph-14.2.21-260-g73cf0b96/src/test/libcephfs/test.cc: In member function 'virtual void LibCephFS_LookupVino_Test::TestBody()':
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/gigantic/release/14.2.21-260-g73cf0b96/rpm/el7/BUILD/ceph-14.2.21-260-g73cf0b96/src/test/libcephfs/test.cc:2498:16: error: 'ceph_mksnap' was not declared in this scope`

see https://shaman.ceph.com/builds/ceph/wip-yuri3-testing-2021-05-21-1132-nautilus/
https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos7,DIST=centos7,MACHINE_SIZE=gigantic/54689//consoleFull

**Comment by smithfarm** [comment]
@yuriw Yup.

File: src/test/libcephfs/test.cc [hunk 1]
```diff
@@ -2469,3 +2470,74 @@ TEST(LibCephFS, Lseek) {
...
+  sprintf(file_path, "%s/file_%d", dir_path, getpid());
+  sprintf(snapdir_path, "%s/.snap", dir_path);
+  sprintf(snap_path, "%s/%s", snapdir_path, snap_name);
+  sprintf(snap_file, "%s/file_%d", snap_path, getpid());
+
+  ASSERT_EQ(0, ceph_mkdir(cmount, dir_path, 0755));
+  int fd = ceph_open(cmount, file_path, O_WRONLY|O_CREAT, 0666);
+  ASSERT_LE(0, fd);
+  ASSERT_EQ(0, ceph_close(cmount, fd));
+  ASSERT_EQ(0, ceph_mksnap(cmount, dir_path, snap_name, 0755, nullptr, 0));
```
Comment (ajarr) [inline src/test/libcephfs/test.cc:2498]: @jtlayton the build is failing here, as mentioned in,
https://github.com/ceph/ceph/pull/40769#issuecomment-846441305

`ceph_mksnap` was introduced in PR https://github.com/ceph/ceph/pull/37721/  The PR doesn't seem simple to backport.  Can the the test be modified to just do a mkdir of a '.snap'  directory explicitly and avoid the need of 'ceph_mksnap`?

Comment (smithfarm) [reply src/test/libcephfs/test.cc:2498]: >  just do a mkdir of a '.snap' directory explicitly

@ajarr Yes, that's what @jtlayton suggested in the discussion we had in the octopus backport (which is identical to this one).

All that is needed is the actual code to replace the `ceph_mksnap` call.

Comment (jtlayton) [reply src/test/libcephfs/test.cc:2507]: Ok, just added a proposed patch. Let's see if it builds...

Comment (ajarr) [reply src/test/libcephfs/test.cc:2507]: @jtlayton @smithfarm thanks!

File: src/test/libcephfs/test.cc [hunk 2]
```diff
@@ -2469,3 +2470,110 @@ TEST(LibCephFS, Lseek) {
...
+  ceph_ll_put(cmount, inode);
+  ASSERT_EQ(0, ceph_ll_lookup_vino(cmount, snapdir_vino, &inode));
+  ceph_ll_put(cmount, inode);
+  ASSERT_EQ(0, ceph_ll_lookup_vino(cmount, snap_vino, &inode));
+  ceph_ll_put(cmount, inode);
+  ASSERT_EQ(0, ceph_ll_lookup_vino(cmount, snap_file_vino, &inode));
+  ceph_ll_put(cmount, inode);
+
+  // cleanup
+  ASSERT_EQ(0, ceph_rmsnap(cmount, dir_path, snap_name));
```
Comment (ajarr) [inline src/test/libcephfs/test.cc:2547]: @jtlayton `ceph_rmsnap` needs definition too.
https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos7,DIST=centos7,MACHINE_SIZE=gigantic/54832/consoleFull#16470259302a811ea2-3e7b-466b-84b4-d13df7e35809

**Comment by ajarr** [comment]
@lxbsz I've not seen this failure before https://pulpito.ceph.com/yuriw-2021-05-27_19:31:33-kcephfs-wip-yuri3-testing-2021-05-27-0818-nautilus-distro-basic-smithi/6138429/
can you please check whether your change in this PR has anything to do this? On first glance, doesn't seem like it, but can you confirm?

**Comment by lxbsz** [comment]
> @lxbsz I've not seen this failure before https://pulpito.ceph.com/yuriw-2021-05-27_19:31:33-kcephfs-wip-yuri3-testing-2021-05-27-0818-nautilus-distro-basic-smithi/6138429/
> can you please check whether your change in this PR has anything to do this? On first glance, doesn't seem like it, but can you confirm?

Checked it, it is not related to any commit in this PR, it was trying to allocate a very large memory `94317937795072`, which triggered oom for `mds.c` daemon:

```
2021-05-27T23:14:13.379 INFO:tasks.ceph.mds.c.smithi104.stderr:tcmalloc: large alloc 94317937795072 bytes == (nil) @  0x7f75fe0ea35f 0x7f75fe109a56 0x55c817ca10f9 0x55c817ca127b 0x55c817ca1340 0x55c817ca141e 0x55c817e90230 0x55c817e904cf 0x55c817d7af8d 0x55c817f33ea4 0x55c817f340ff 0x55c817f34310 0x7f75fe7b151f 0x7f75fc5dbea5 0x7f75fb2889fd
2021-05-27T23:14:13.379 INFO:tasks.ceph.mds.c.smithi104.stderr:terminate called after throwing an instance of 'std::bad_alloc'
2021-05-27T23:14:13.379 INFO:tasks.ceph.mds.c.smithi104.stderr:  what():  std::bad_alloc
2021-05-27T23:14:13.380 INFO:tasks.ceph.mds.c.smithi104.stderr:*** Caught signal (Aborted) **
2021-05-27T23:14:13.380 INFO:tasks.ceph.mds.c.smithi104.stderr: in thread 7f75eec84700 thread_name:fn_anonymous
2021-05-27T23:14:13.381 INFO:tasks.ceph.mds.c.smithi104.stderr: ceph version 14.2.21-281-gb7a4754d941 (b7a4754d941f016678ca17d548fafb10e7444735) nautilus (stable)
2021-05-27T23:14:13.381 INFO:tasks.ceph.mds.c.smithi104.stderr: 1: (()+0xf630) [0x7f75fc5e3630]
2021-05-27T23:14:13.381 INFO:tasks.ceph.mds.c.smithi104.stderr: 2: (gsignal()+0x37) [0x7f75fb1c03d7]
2021-05-27T23:14:13.382 INFO:tasks.ceph.mds.c.smithi104.stderr: 3: (abort()+0x148) [0x7f75fb1c1ac8]
2021-05-27T23:14:13.382 INFO:tasks.ceph.mds.c.smithi104.stderr: 4: (__gnu_cxx::__verbose_terminate_handler()+0x165) [0x7f75fbad0a95]
2021-05-27T23:14:13.382 INFO:tasks.ceph.mds.c.smithi104.stderr: 5: (()+0x5ea06) [0x7f75fbacea06]
2021-05-27T23:14:13.382 INFO:tasks.ceph.mds.c.smithi104.stderr: 6: (()+0x5ea33) [0x7f75fbacea33]
2021-05-27T23:14:13.383 INFO:tasks.ceph.mds.c.smithi104.stderr: 7: (()+0x5ec53) [0x7f75fbacec53]
2021-05-27T23:14:13.383 INFO:tasks.ceph.mds.c.smithi104.stderr: 8: (()+0x1772b) [0x7f75fe0eb72b]
2021-05-27T23:14:13.383 INFO:tasks.ceph.mds.c.smithi104.stderr: 9: (tcmalloc::allocate_full_cpp_throw_oom(unsigned long)+0xf3) [0x7f75fe109a03]
2021-05-27T23:14:13.383 INFO:tasks.ceph.mds.c.smithi104.stderr: 10: (std::string::_Rep::_S_create(unsigned long, unsigned long, std::allocator<char> const&)+0x59) [0x55c817ca10f9]
2021-05-27T23:14:13.383 INFO:tasks.ceph.mds.c.smithi104.stderr: 11: (std::string::_Rep::_M_clone(std::allocator<char> const&, unsigned long)+0x1b) [0x55c817ca127b]
2021-05-27T23:14:13.384 INFO:tasks.ceph.mds.c.smithi104.stderr: 12: (std::string::reserve(unsigned long)+0x40) [0x55c817ca1340]
2021-05-27T23:14:13.384 INFO:tasks.ceph.mds.c.smithi104.stderr: 13: (std::string::append(char const*, unsigned long)+0x5e) [0x55c817ca141e]
2021-05-27T23:14:13.384 INFO:tasks.ceph.mds.c.smithi104.stderr: 14: (CDentry::make_path(filepath&, bool) const+0x90) [0x55c817e90230]
2021-05-27T23:14:13.384 INFO:tasks.ceph.mds.c.smithi104.stderr: 15: (operator<<(std::ostream&, CDentry const&)+0x6f) [0x55c817e904cf]
2021-05-27T23:14:13.385 INFO:tasks.ceph.mds.c.smithi104.stderr: 16: (MDCache::_create_system_file_finish(boost::intrusive_ptr<MutationImpl>&, CDentry*, unsigned long, MDSContext*)+0xbd) [0x55c817d7af8d]
2021-05-27T23:14:13.385 INFO:tasks.ceph.mds.c.smithi104.stderr: 17: (MDSContext::complete(int)+0x74) [0x55c817f33ea4]
2021-05-27T23:14:13.385 INFO:tasks.ceph.mds.c.smithi104.stderr: 18: (MDSIOContextBase::complete(int)+0x16f) [0x55c817f340ff]
2021-05-27T23:14:13.385 INFO:tasks.ceph.mds.c.smithi104.stderr: 19: (MDSLogContextBase::complete(int)+0x40) [0x55c817f34310]
2021-05-27T23:14:13.385 INFO:tasks.ceph.mds.c.smithi104.stderr: 20: (Finisher::finisher_thread_entry()+0x16f) [0x7f75fe7b151f]
2021-05-27T23:14:13.386 INFO:tasks.ceph.mds.c.smithi104.stderr: 21: (()+0x7ea5) [0x7f75fc5dbea5]
2021-05-27T23:14:13.386 INFO:tasks.ceph.mds.c.smithi104.stderr: 22: (clone()+0x6d) [0x7f75fb2889fd]
2021-05-27T23:14:13.386 INFO:tasks.ceph.mds.c.smithi104.stderr:2021-05-27 23:14:13.415 7f75eec84700 -1 *** Caught signal (Aborted) **
2021-05-27T23:14:13.386 INFO:tasks.ceph.mds.c.smithi104.stderr: in thread 7f75eec84700 thread_name:fn_anonymous
2021-05-27T23:14:13.387 INFO:tasks.ceph.mds.c.smithi104.stderr: ceph version 14.2.21-281-gb7a4754d941 (b7a4754d941f016678ca17d548fafb10e7444735) nautilus (stable)
2021-05-27T23:14:13.387 INFO:tasks.ceph.mds.c.smithi104.stderr: 1: (()+0xf630) [0x7f75fc5e3630]
2021-05-27T23:14:13.387 INFO:tasks.ceph.mds.c.smithi104.stderr: 2: (gsignal()+0x37) [0x7f75fb1c03d7]
2021-05-27T23:14:13.387 INFO:tasks.ceph.mds.c.smithi104.stderr: 3: (abort()+0x148) [0x7f75fb1c1ac8]
2021-05-27T23:14:13.388 INFO:tasks.ceph.mds.c.smithi104.stderr: 4: (__gnu_cxx::__verbose_terminate_handler()+0x165) [0x7f75fbad0a95]
2021-05-27T23:14:13.388 INFO:tasks.ceph.mds.c.smithi104.stderr: 5: (()+0x5ea06) [0x7f75fbacea06]
2021-05-27T23:14:13.388 INFO:tasks.ceph.mds.c.smithi104.stderr: 6: (()+0x5ea33) [0x7f75fbacea33]
2021-05-27T23:14:13.388 INFO:tasks.ceph.mds.c.smithi104.stderr: 7: (()+0x5ec53) [0x7f75fbacec53]
2021-05-27T23:14:13.389 INFO:tasks.ceph.mds.c.smithi104.stderr: 8: (()+0x1772b) [0x7f75fe0eb72b]
2021-05-27T23:14:13.389 INFO:tasks.ceph.mds.c.smithi104.stderr: 9: (tcmalloc::allocate_full_cpp_throw_oom(unsigned long)+0xf3) [0x7f75fe109a03]
2021-05-27T23:14:13.389 INFO:tasks.ceph.mds.c.smithi104.stderr: 10: (std::string::_Rep::_S_create(unsigned long, unsigned long, std::allocator<char> const&)+0x59) [0x55c817ca10f9]
2021-05-27T23:14:13.389 INFO:tasks.ceph.mds.c.smithi104.stderr: 11: (std::string::_Rep::_M_clone(std::allocator<char> const&, unsigned long)+0x1b) [0x55c817ca127b]
2021-05-27T23:14:13.389 INFO:tasks.ceph.mds.c.smithi104.stderr: 12: (std::string::reserve(unsigned long)+0x40) [0x55c817ca1340]
2021-05-27T23:14:13.390 INFO:tasks.ceph.mds.c.smithi104.stderr: 13: (std::string::append(char const*, unsigned long)+0x5e) [0x55c817ca141e]
2021-05-27T23:14:13.390 INFO:tasks.ceph.mds.c.smithi104.stderr: 14: (CDentry::make_path(filepath&, bool) const+0x90) [0x55c817e90230]
2021-05-27T23:14:13.390 INFO:tasks.ceph.mds.c.smithi104.stderr: 15: (operator<<(std::ostream&, CDentry const&)+0x6f) [0x55c817e904cf]
2021-05-27T23:14:13.390 INFO:tasks.ceph.mds.c.smithi104.stderr: 16: (MDCache::_create_system_file_finish(boost::intrusive_ptr<MutationImpl>&, CDentry*, unsigned long, MDSContext*)+0xbd) [0x55c817d7af8d]
2021-05-27T23:14:13.391 INFO:tasks.ceph.mds.c.smithi104.stderr: 17: (MDSContext::complete(int)+0x74) [0x55c817f33ea4]
2021-05-27T23:14:13.391 INFO:tasks.ceph.mds.c.smithi104.stderr: 18: (MDSIOContextBase::complete(int)+0x16f) [0x55c817f340ff]
2021-05-27T23:14:13.391 INFO:tasks.ceph.mds.c.smithi104.stderr: 19: (MDSLogContextBase::complete(int)+0x40) [0x55c817f34310]
2021-05-27T23:14:13.391 INFO:tasks.ceph.mds.c.smithi104.stderr: 20: (Finisher::finisher_thread_entry()+0x16f) [0x7f75fe7b151f]
2021-05-27T23:14:13.391 INFO:tasks.ceph.mds.c.smithi104.stderr: 21: (()+0x7ea5) [0x7f75fc5dbea5]
2021-05-27T23:14:13.392 INFO:tasks.ceph.mds.c.smithi104.stderr: 22: (clone()+0x6d) [0x7f75fb2889fd]
2021-05-27T23:14:13.392 INFO:tasks.ceph.mds.c.smithi104.stderr: NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.
2021-05-27T23:14:13.392 INFO:tasks.ceph.mds.c.smithi104.stderr:
2021-05-27T23:14:13.392 INFO:tasks.ceph.mds.c.smithi104.stderr:  -405> 2021-05-27 23:14:10.413 7f75eec84700 -1 mds.0.openfiles _load_finish got (2) No such file or directory
2021-05-27T23:14:13.393 INFO:tasks.ceph.mds.c.smithi104.stderr:     0> 2021-05-27 23:14:13.415 7f75eec84700 -1 *** Caught signal (Aborted) **
2021-05-27T23:14:13.393 INFO:tasks.ceph.mds.c.smithi104.stderr: in thread 7f75eec84700 thread_name:fn_anonymous
2021-05-27T23:14:13.393 INFO:tasks.ceph.mds.c.smithi104.stderr:
2021-05-27T23:14:13.393 INFO:tasks.ceph.mds.c.smithi104.stderr: ceph version 14.2.21-281-gb7a4754d941 (b7a4754d941f016678ca17d548fafb10e7444735) nautilus (stable)

```
I think the master also have the same problem, I have raised on tracker: https://tracker.ceph.com/issues/51023 to follow this.

**Review by ajarr — APPROVED** [review]
@lxbsz thanks a lot!

@yuriw please merge this PR

**Pull request merged**